### PR TITLE
add lru pool to reduce memory usage when #feature is large(#4)

### PR DESCRIPTION
**Pull request by guolinke**
To slove #4

### CODE REVIEW

File: include/LightGBM/config.h [hunk 1]
```diff
@@ -140,6 +140,8 @@ struct TreeConfig: public ConfigBase {
   int num_leaves = 127;
   int feature_fraction_seed = 2;
   double feature_fraction = 1.0;
+  // max cache size(unit:GB) for historical histogram. < 0 means not limit
```
Comment (chivee) [inline include/LightGBM/config.h]: I think it will better to using MB as unit size

File: src/io/config.cpp [hunk 2]
```diff
@@ -110,28 +110,30 @@ void OverallConfig::GetTaskType(const std::unordered_map<std::string, std::strin
...
-                                                 TreeLearnerType::kFeatureParallelTreelearner) {
+  if (gbdt_config->tree_learner_type == TreeLearnerType::kSerialTreeLearner ||
+    gbdt_config->tree_learner_type == TreeLearnerType::kFeatureParallelTreelearner) {
     is_parallel_find_bin = false;
-  } else if (dynamic_cast<GBDTConfig*>(boosting_config)->tree_learner_type ==
-                                                 TreeLearnerType::kDataParallelTreeLearner) {
+  } else if (gbdt_config->tree_learner_type == TreeLearnerType::kDataParallelTreeLearner) {
     is_parallel_find_bin = true;
+    if (gbdt_config->tree_config.histogram_pool_size >= 0) {
+      Log::Error("Auto set histogram_pool_size to non-limit, to reduce communication cost");
```
Comment (chivee) [inline src/io/config.cpp]: " Histogram LRU queue was enabled for reducing communication cost. ( histogram_pool_size = %f )", tree_config.histogram_pool_size

File: src/io/config.cpp [hunk 3]
```diff
@@ -110,28 +110,30 @@ void OverallConfig::GetTaskType(const std::unordered_map<std::string, std::strin
...
+      // To reduce communication cost, not limit pool size when using data parallel
```
Comment (chivee) [inline src/io/config.cpp]: Change pool size to -1 when using data parallel for reducing communication cost

File: src/treelearner/serial_tree_learner.cpp [hunk 4]
```diff
@@ -46,16 +46,32 @@ void SerialTreeLearner::Init(const Dataset* train_data) {
...
+    max_cache_size = num_leaves_;
+  } else {
+    size_t total_histogram_size = 0;
+    for (int i = 0; i < train_data_->num_features(); ++i) {
+      total_histogram_size += sizeof(HistogramBinEntry) * train_data_->FeatureAt(i)->num_bin();
+    }
+    max_cache_size = static_cast<int>(histogram_pool_size_ * 1024 * 1024 * 1024 / total_histogram_size);
+    
+  }
+  // at least need 2 leaves
```
Comment (chivee) [inline src/treelearner/serial_tree_learner.cpp]: here is assuming the num_leaves couldn't less than 2, right?

File: src/io/config.cpp (hunk 3 above)
Comment (guolinke) [inline src/io/config.cpp]: Do you means alter the output and comment?

File: include/LightGBM/config.h (hunk 1 above)
Comment (guolinke) [inline include/LightGBM/config.h]: I think both MB and GB are OK for this is a float parameter.

Comment (chivee) [inline include/LightGBM/config.h]: yes, but considering this option will be used only on machine with limited memories, offer an    fine-grained option will be nice. you can multiply 1024 to make code unchanged

**Pull request merged**